pkg/kubelet/dockershim/exec.go

```diff
@@ -74,62 +79,93 @@ func (*NativeExecHandler) ExecInContainer(client libdocker.Interface, container
+		// exec has started; otherwise, the initial resize request will fail.
+		execStarted := make(chan struct{})
+		go func() {
+			select {
+			case <-execStarted:
+				// client.StartExec has started the exec, so we can start resizing
+			case <-done:
+				// ExecInContainer has returned, so short-circuit
+				return
+			}
 
-		kubecontainer.HandleResizing(resize, func(size remotecommand.TerminalSize) {
-			client.ResizeExecTTY(execObj.ID, uint(size.Height), uint(size.Width))
-		})
-	}()
+			kubecontainer.HandleResizing(resize, func(size remotecommand.TerminalSize) {
+				client.ResizeExecTTY(execObj.ID, uint(size.Height), uint(size.Width))
+			})
+		}()
+
+		startOpts := dockertypes.ExecStartCheck{Detach: false, Tty: tty}
+		streamOpts := libdocker.StreamOptions{
+			InputStream:  stdin,
+			OutputStream: stdout,
+			ErrorStream:  stderr,
+			RawTerminal:  tty,
+			ExecStarted:  execStarted,
+		}
+		err = client.StartExec(execObj.ID, startOpts, streamOpts)
+		if err != nil {
+			return err
+		}
 
-	startOpts := dockertypes.ExecStartCheck{Detach: false, Tty: tty}
-	streamOpts := libdocker.StreamOptions{
-		InputStream:  stdin,
-		OutputStream: stdout,
-		ErrorStream:  stderr,
-		RawTerminal:  tty,
-		ExecStarted:  execStarted,
-	}
-	err = client.StartExec(execObj.ID, startOpts, streamOpts)
-	if err != nil {
-		return err
-	}
+		ticker := time.NewTicker(2 * time.Second)
+		defer ticker.Stop()
+		count := 0
+		for {
+			inspect, err2 := client.InspectExec(execObj.ID)
+			if err2 != nil {
+				return err2
+			}
+			if !inspect.Running {
+				if inspect.ExitCode != 0 {
+					err = &dockerExitError{inspect}
+				}
+				break
+			}
 
-	ticker := time.NewTicker(2 * time.Second)
-	defer ticker.Stop()
-	count := 0
-	for {
-		inspect, err2 := client.InspectExec(execObj.ID)
-		if err2 != nil {
-			return err2
-		}
-		if !inspect.Running {
-			if inspect.ExitCode != 0 {
-				err = &dockerExitError{inspect}
+			count++
+			if count == 5 {
+				klog.Errorf("Exec session %s in container %s terminated but process still running!", execObj.ID, container.ID)
+				break
 			}
-			break
-		}
 
-		count++
-		if count == 5 {
-			klog.Errorf("Exec session %s in container %s terminated but process still running!", execObj.ID, container.ID)
-			break
+			<-ticker.C
+		}
+		return err
+	}
+	// No timeout, block until startExec is finished.
+	if timeout <= 0 {
+		return startExec()
+	}
+	// Otherwise, run startExec in a new goroutine and wait for completion
+	// or timeout, whatever happens first.
+	ch := make(chan error, 1)
+	go func() {
+		ch <- startExec()
+	}()
+	select {
+	case err := <-ch:
+		return err
+	case <-time.After(timeout):
+		// FIXME: we should kill the process in the container, but the
+		// Docker API doesn't support it. See
+		// https://github.com/docker/docker/issues/9098.
+		// For liveness probes this is probably okay, since the
+		// container will be restarted. For readiness probes it means
```

Comment:
I'm a little worried about this behavior... specifically in the following situation:

Imagine we have a readiness probe configured to run every minute. The timeout is set to 5s, but the actual command we are executing never terminates.

Because we can't kill the process in the container, I believe we're going to create more and more processes in the container (and more and more goroutines), which seems dangerous to me?

Reply:
I think this was where the previous PR got stalled.

I left a comment on the linked moby issue - surki updated his branch in Dec 2019. Hopefully that issue can get some traction.

For the readiness probe you mentioned, maybe we can adopt a threshold for the timeout so that we don't create many processes.

Reply:
@dims thoughts?

I'm curious... is the lack of timeout in the `ExecInContainer` causing issues for users in production? I'm just wondering if doing nothing until this is fixed in `moby` is a feasible option...